text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+person.lastName = 'Jackson';
+```
+
+Now `fullName` knows which tracked properties were accessed when it was run,
+and setting `lastName` has invalidated `fullName`.
+
+> **NOTE:** This does _not_ invalidate a cache like in computed properties. Even
+> if `firstName` and `lastName` were untracked, the tracked getter would still
+> return the correct value on subsequent accesses, because `@tracked` does _not_
+> cache values. The validation and invalidation is pure meta data that is only
+> accessible by the Glimmer VM.
+>
+> Internally, Glimmer checks to see if a value has updated _before calling the
+> getter_. If it hasn't, then Glimmer does not rerender the related section of
+> the DOM. This is effectively an automatic `shouldComponentUpdate` from React.
+>
+> To prevent inconsistency, during development time, tracked properties will
+> keep a cache of their previous value to compare when they are activated and
+> ensure that it hasn't changed without invalidation. This will prevent improper
+> usage of tracked properties _outside_ of Glimmer's change tracking.
+
+### Manual Invalidation
+
+In user code, the idea that all mutable properties should be marked as tracked
+and that all other properties are effectively immutable works well in isolation.
+However, there are cases where users will want to work with code they do _not_
+control, such as external library code.
+
+Consider the following example. We have a `simple-timer` library that we've
+imported from NPM, and we're trying to wrap it with a `TimerComponent` that
+uses it to keep track of how much time has passed:
+
+```js
+// simple-timer/index.js
+export default class Timer {
+  seconds = 0;
+  minutes = 0;
+  hours = 0;
+
+  listeners = [];
+
+  constructor() {
+    setInterval(() => {
+      this.seconds++;
+      this.minutes = Math.floor(this.seconds / 60);
+      this.hours = Math.floor(this.minutes / 60);
+      this.notifyTick();
+    }, 1000);
+  }
+
+  notifyTick() {
+    for (let listener of this.listeners) {
+      listener(this.seconds);
+    }
+  }
+
+  onTick(listener) {
+    this.listeners.push(listener);
+  }
+}
+```
+
+```js
+import Timer from 'simple-timer';
+import Component, { tracked } from '@glimmer/tracking';
+
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+Even though we've marked the `timer` property as tracked, the `timer.seconds`
+property is untracked, and _it_ is the field that is updated. We can solve this
+problem by using the timer library's `onTick` event handler to re-set the field,
+invalidating it:
+
+```js
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  constructor() {
+    this.timer.onTick(() => {
+      // invalidate the timer field.
+      this.timer = this.timer;
```

Comment:
this is very confusing. I understand what it does, but to anyone unfamiliar, this will look like a no-op that can be deleted. I would prefer it if we could find a more descriptive way to achieve the same thing.

Reply:
Note that this is the semantic equivalent to `this.setState({ timer })` in React to trigger a revalidation. We have some ideas for how we might make this more explicit in the future (as well as potentially eke out some small performance optimizations) but wanted to decouple that from this RFC.

Reply:
I agree with @Turbo87 that, while this makes sense when you understand how tracked properties work under the hood, it will confuse many Ember users. +1 for a more explicit way to mark a tracked property as dirty.

Reply:
Long term, I definitely agree. I actually think we should expose Glimmer's references and validators APIs in a way that allows libraries to intelligently instrument themselves, more so than what we have with tracked properties.

That may take a while though. We bikeshedded for quite a while on some intermediate solutions, including integrating with `notifyPropertyChange` (avoided so tracked props aren't dependent on Ember) or something like it, but landed on this because it is the _minimal_ proposal. It works because of the nature of tracked properties, and it would work _even with alternatives_. It allows us to spend some time figuring out what a better alternative would be.

Also, FWIW, a very simple notify function implementable in user-land would be:

```js
function notify(obj, key) {
  obj[key] = obj[key];
}
```

Reply:
I like this, but not sure about the name. "Whom are we notifying and what are we notifying them of?", says the casual user. If this will be a standard export of `@ember/tracking`, something like `markDirty()` or `hasChanged()` or `updated()` seems better. The docs should also make it clear that this is only necessary for changes to object properties/array elements (and perhaps it should even debug assert when used on non-objects to prevent unnecessary invalidation.)

Reply:
I agree that it would need another name. My initial idea was an `invalidate()` method, but `markDirty()` works for me as well.

Next question is how do you prevent people from using `Ember.get()` nested path syntax for the key name? I'd hate if it had to check all keys for a `"."`.

Reply:
> Next question is how do you prevent people from using `Ember.get()` nested path syntax for the key name? I'd hate if it had to check all keys for a `"."`.

maybe in a separate RFC, we could discuss a deprecation strategy `Ember.get`?

Reply:
I definitely agree a function that calls out that you're intending to update a Tracked Property is  needed.  `this.foo = this.foo` has no discernible intent programmatically, and is only being used for its side effect. To put that code in somewhere requires knowing quite a lot not just about the inner workings of tracked properties, but also the specifics of the business logic in that particular corner of your code, because in general it won't be needed.

How about as an option for a method name:
 ```js
trackChange(obj, prop)
```

Btw, I LOVE this proposal! Thank you!

Reply:
Looking at this RFC, I only see a reference to reducing the API surface area in the summary, and not mentioned in the motivation section -- maybe this is a shortcoming in the way the RFC is written, but as written, that doesn't look like a primary motivation, so I've mentally discarded that.

So the main RFC motivation to which the `this.timer = this.timer` question applies is `Leveraging Existing JavaScript knowledge`. And I think this proposal (almost) totally nails that goal. A huge part of why I like the proposal so much is that it strikes just the right balance of making the code look vanilla and Javascript-y enough to allow users to read it without having or leveraging a bunch of framework knowledge, but deviates from vanilla Javascript enough to remind the reader of the fact that more is happening than vanilla Javascript so they can decide whether to read the code with the framework feature mentally encapsulated (native Javascript with some extra tooling that I'm treating as a black box and not really thinking about) or with the framework feature exposed and integrated into their reading of the code, with the commensurate (and sometimes, but only sometimes, necessary) increase in mental load.

I think enabling that choice is huge for ergonomics, learnability, and maintainability, because it helps new users progress from encapsulated thinking to un-encapsulated thinking, learning iteratively and hopefully being less intimidated, and also for experienced users because it's very helpful to be able to decide when it's necessary to break the mental encapsulation of the framework feature, and when to avoid that extra mental load.

I think `this.timer = this.timer` runs counter to that. For experienced developers, it would rapidly turn into the same sort of indicator of "magic lives here" as `@tracked`. It might be a little less friendly (scanning code it wouldn't as easily jump out at me as something framework-related), but probably not very impactfully so. But for new-to-Ember developers, I think it runs entirely counter to the `Leverage Existing JavaScript Knowledge`. It would be leveraging it, but leveraging it in the service of confusion rather than learning. It would be leveraging it to actively hide the "magic lives here" learning cue, especially because of how un-google-able it is.

Implementing some kind of `invalidate(obj, key)` would allow for this encapsulation choice -- immediately clear that magic lives here, something google-able to help me learn if needed, and a reminder/opportunity to stop reading with the simpler vanilla-Javascript model and start adding tracked property complexity to my mental model, if needed for what I'm doing.

I really think this is a case where making it look more like vanilla Javascript serves to confuse and trick, rather than clarify and make familiar.

Reply:
Having `this.timer = this.timer` being the proposed way to invalidate stuff, will add similar confusion as needing to get `this.timer` once to initialize observers depending on it. I think we all faced the confusion that one caused - and wasted a lot of time into debugging issues caused by it. So please don't add something like this again.

Reply:
I would much prefer that this side-effect pattern is not usable at all. Ie. that `this.timer = this.timer` does not trigger any observers.

I guess that this could be done by calling the getter in the setter, and do a simple `===` comparison before invalidating it?

Unfortunately, this will mean that either a custom setter ala `set()` must be used, or that the handler caches the old value to compare and know when it has changed.

Reply:
@kanongil Unfortunately we would have to cache the value of the getter in order to do that, and getting rid of cache-by-default is a major goal of tracked properties. Setting up the cache itself is fairly non-performant as is, and while this _could_ be mitigated with tracked properties to some extent (since we know what properties would be cached ahead of time), we definitely can't get around the fact that it will cause us to hold onto extra references and take up extra memory, which can become a problem in larger and longer lived applications.

Reply:
@pzuraq See my getter proposal in my comment. It won't require any caching.

I'd say that removing the side-effect, and making the invalidation explicit, is a better way to introduce this feature. If it causes too many issues, I guess the side-effect can be re-introduced at a later stage.

Reply:
@kanongil it requires that we either cache the value, or rerun the getter at the time of setting. Part of the benefit of the current setup, and pull-based change tracking in general, is that no matter how many times you update a value, you only do the expensive work once:

```js
@tracked 
get arr() {
	return this._arr.sort();
}

set arr(value) {
	this._arr = value;
}
```

This is not a great example because there are other ways you could set this up, but you can imagine how if the array was 10,000 items, the sort could be fairly expensive. Now lets imagine that we set this array several times in the course of a single render:

```js
this.arr.push(...items);
this.arr = this.arr.slice();

// ...

this.arr.push(...items);
this.arr = this.arr.slice();

// ...

this.arr.push(...items);
this.arr = this.arr.slice();
```

All of this happens sychronously. Every time we do that check for array equality, we are running the sort in the getter. We run it 3 times when updating the items, and then Glimmer runs it one more time when rendering them.

Currently, whenever `this.arr` is set, the tracked property bumps a number counter. This is _much_ cheaper in almost all situations, probably one of the cheapest things we can do in the VM. While it's behavior may seem a bit confusing at first, it also means we can't accidentally do more work than we expected. Each function runs at most once.

Reply:
@kanongil I really think we have to make `this.timer = this.timer` invalidate. If that doesn't invalidate, then a case where another object gets this object's tracked value (array/object), mutates it, and then sets it again would not invalidate, and that is going to be really really confusing and hard to reason about/debug.

I think we need to keep the simple rule of "if you set it, it's invalidated" always true, or we're going to open the door for a whole bunch use cases that cause unexpected behavior and hard-to-find bugs.

Reply:
> I think we need to keep the simple rule of "if you set it, it's invalidated" always true, or we're going to open the door for a whole bunch use cases that cause unexpected behavior and hard-to-find bugs.

Agreed. Though I don't think that this means we can't _also_ have an additional `invalidate` / `notify` method...

Reply:
> Currently, whenever this.arr is set, the tracked property bumps a number counter. This is much cheaper in almost all situations, probably one of the cheapest things we can do in the VM. While it's behavior may seem a bit confusing at first, it also means we can't accidentally do more work than we expected. Each function runs at most once.

This is a nice property of the system but only applies to the actual rendering pass. If used outside rendering, it will re-run the expensive getter. E.g. if the property is used in response to incoming events from a remote server.

As such, I think it works better if tracked properties are modelled as "fast", enabling a cheap test to avoid redundant invalidations.

This should also provide a better developer experience, when doing performance optimisations. An expensive getter should show up in profiles, and could often be fixed by applying local caching. This is much nicer than diagnosing the unbounded redundant invalidations the proposed system contains.

Note that this change also makes it _more compatible_ with the listed motivations.

Reply:
Actually, the "we can't accidentally do more work than we expected" assertion doesn't even hold for glimmer usage. Say another `@tracked` property was added like this:

```js
@tracked
get currentTimeString() {
  return `${this.timer.minutes}:${this.timer.minutes}`;
}
```

Here the `timer` getter would be called 2 times, though it doesn't change in-between. This illustrates that the proposed system requires careful consumption of properties to avoid inefficiencies.

Reply:
I think something I haven't really communicated very well here is that there is no silver bullet when it comes to performance. In some cases, it is cheaper overall to rerun getters slightly more often than to hold onto large amounts of memory by eagerly caching everything. In others, caching is definitely more performant, because the work done by the getters is fairly expensive.

Currently, with computed properties, we don't have any granularity around this in Ember. Everything is cached, and users can't opt out of that behavior. We proposed `@tracked` as a simpler change tracking layer partially to enable that choice, and partially because we believe it optimizes for the common case (relatively cheap getters).

One additional feature we want to add is an `@memo` decorator, which would allow users to opt back into caching:

```js
@memo
@tracked
get fullName() {}
```

We decided not to include it in this RFC because it opens a lot of room around bikeshedding the behavior (should it be closer to computed? Should it always trigger setters? Should it require both decorators?) and because we were hoping to build it when the underlying primitives were made public, but it would be pretty simple to build and we wanted to propose it in the near future. 

@kanongil do you think this additional decorator would address your concerns? And if so, do you think it should be added to this proposal?

Reply:
This comment thread has a few sub-threads going now, and I think each of them merits its own discussion and thought. But, while the concerns raised are real, I want to reiterate how rare this situation will be in the vast majority of apps.

Remember: this only comes up when interacting with objects from third-party libraries, where those objects are mutated by the library over time _and you want to rely on those properties directly from your templates_.

This is actually not supported at all in Ember today. Why? Because if you use an external POJO in a template or computed property, Ember will install the "mandatory setter" that causes an exception to be thrown when the library mutates its own object.

In practice, this ends up being pretty rare in app code. Reflecting changes from a third-party library into something your UI framework understands (whether it's `setState`, `Ember.set` or something else) is a problem ecosystem-wide, and we've developed strategies for dealing with it. Usually there will be an addon that adapts a popular library to the idioms of your particular library/framework, which is likely to still be true here.

Because of how widespread this problem is, it also means that most libraries expose callback APIs that pass changed properties as arguments, allowing you to reflect the most up-to-date state onto your component directly.

Even though the RFC shows a case where you _can_ directly utilize mutable, untracked properties on a POJO, I think a more common factoring will be to simply copy properties to the component as they change:

```js
export default class TimerComponent extends Component {
  timer = new Timer();

  @tracked seconds;
  @tracked minutes;

  constructor() {
    this.timer.onTick(({ seconds, minutes }) => {
      this.seconds = seconds;
      this.minutes = minutes;
    });
  }
}
```

In this example, the timer's `onTick` callback passes the updated `seconds` and `minutes` values. Not only is this factoring less confusing (avoiding the ambiguous `this.timer = this.timer`), it's also less code overall. 😁 

Given the rarity of encountering this situation, and given that there are clear ways of refactoring to avoid the problem, I personally don't think it rises to the level of being worth introducing and teaching an entirely new API.